src/Education/XamlGenerator/RuleHelpXamlTranslator.cs

```diff
@@ -0,0 +1,454 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
+using static Microsoft.VisualStudio.Threading.AsyncReaderWriterLock;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Xml;
+using System.Collections.Generic;
+using System;
+
+namespace SonarLint.VisualStudio.Education.XamlGenerator
+{
+    public interface IRuleHelpXamlTranslator
+    {
+        string TranslateHtmlToXaml(string htmlContent, bool supportsInline = false);
+    }
+
+    internal class RuleHelpXamlTranslator : IRuleHelpXamlTranslator
+    {
+        private XmlWriter writer;
+        private XmlReader reader;
+
+        /// <summary>
+        /// Stack of currently open XAML elements
+        /// </summary>
+        /// <remarks>We need some information about the current structure so we can check whether some 
+        /// operations are valid e.g. can be we add text to the current element?</remarks>
+        private Stack<XamlOutputElementInfo> outputXamlElementStack;
+
+        /// <summary>
+        /// Used to add background colour to alternate rows in a table
+        /// </summary>
+        private bool tableAlternateRow;
+
+        public string TranslateHtmlToXaml(string htmlContent, bool supportsInline = false)
+        {
+            var sb = new StringBuilder();
+            writer = CreateXmlWriter(sb);
```

Comment:
Out of interest, why are you creating an XML writer here, instead of passing one it?
We're only producing one document, so the same writer could be used throughout.